src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+        return True
+
+
+async def rclone_check_progress(job, proc):
+    RE_TRANSF = re.compile(r'Transferred:\s*?(.+)$', re.S)
+    while True:
+        read = (await proc.stdout.readline()).decode()
+        job.logs_fd.write(read.encode("utf-8", "ignore"))
+        if read == '':
+            break
+        reg = RE_TRANSF.search(read)
+        if reg:
+            transferred = reg.group(1).strip()
+            if not transferred.isdigit():
+                job.set_progress(None, transferred)
+
+
+def rclone_encrypt_password(password):
+    key = bytes([0x9c, 0x93, 0x5b, 0x48, 0x73, 0x0a, 0x55, 0x4d,
+                 0x6b, 0xfd, 0x7c, 0x63, 0xc8, 0x86, 0xa9, 0x2b,
+                 0xd3, 0x90, 0x19, 0x8e, 0xb8, 0x12, 0x8a, 0xfb,
+                 0xf4, 0xde, 0x16, 0x2b, 0x8b, 0x95, 0xf6, 0x38])
+
+    iv = Random.new().read(AES.block_size)
+    counter = Counter.new(128, initial_value=int(codecs.encode(iv, "hex"), 16))
+    cipher = AES.new(key, AES.MODE_CTR, counter=counter)
+    encrypted = iv + cipher.encrypt(password.encode("utf-8"))
+    return base64.urlsafe_b64encode(encrypted).decode("ascii").rstrip("=")
+
+
+def validate_attributes(schema, data, additional_attrs=False):
+    verrors = ValidationErrors()
+
+    schema = Dict("attributes", *schema, additional_attrs=additional_attrs)
+
+    try:
+        data["attributes"] = schema.clean(data["attributes"])
+    except Error as e:
+        verrors.add(e.attribute, e.errmsg, e.errno)
+
+    try:
+        schema.validate(data["attributes"])
+    except ValidationErrors as e:
+        verrors.extend(e)
+
+    return verrors
+
+
+class CredentialsService(CRUDService):
+
+    class Config:
+        namespace = 'cloudsync.credentials'
+
+    @filterable
+    async def query(self, filters=None, options=None):
+        return await self.middleware.call('datastore.query', 'system.cloudcredentials', filters, options)
+
+    @accepts(Dict(
+        'cloud_sync_credentials',
+        Str('name'),
+        Str('provider'),
+        Dict('attributes', additional_attrs=True),
+        register=True,
+    ))
+    async def do_create(self, data):
+        self._validate("cloud_sync_credentials", data)
+
+        data["id"] = await self.middleware.call(
+            'datastore.insert',
+            'system.cloudcredentials',
+            data,
+        )
+        return data
+
+    @accepts(Int('id'), Ref('cloud_sync_credentials'))
+    async def do_update(self, id, data):
+        self._validate("cloud_sync_credentials", data)
+
+        await self.middleware.call(
+            'datastore.update',
+            'system.cloudcredentials',
+            id,
+            data,
+        )
+
+        return data
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        await self.middleware.call(
+            'datastore.delete',
+            'system.cloudcredentials',
+            id,
+        )
+
+    def _validate(self, schema_name, data):
+        provider = REMOTES[data["provider"]]
```

Comment:
Are there guarantees data["provider"] exists in REMOTES? Sounds like a possible KeyError